### ISSUE TITLE
DietPi-Software | Mopidy playlists not working due to wrong config/missing permission

### ISSUE THREAD

**Issue by cyberlussi** [MichaIng/DietPi#2384]
I am new to dietpi and mopidy but it seems for me that  playlists are not working in the vanilla dietpi setup of mopidy. I have a workaround (see below) but I am not sure whether this is perfect under all circumstances. I mean, I do not know whether mopidy-m3u is then still able to scan for m3u below the music directory or not ...

#### Required Information:
- DietPi version	| 6.19.7
- Distro version	| 9.6
- Kernel version	| inux DietPi 4.14.79+ # 1159 Sun Nov 4 17:28:08 GMT 2018 armv6l GNU/Linux
- SBC device		| RPi3
- Power supply used	| 5V 1A
- SDcard used		| Kingston 8GB

#### Additional Information (if applicable):
- Software title	| Mopidy
- Was the software title installed freshly or updated/migrated? YES
- Can this issue be replicated on a fresh installation of DietPi? I think so ...
- `dietpi-bugreport` ID

#### Steps to reproduce:
1. Install with dietpi-software vanilla mopidy
2. Reboot
3. Open mopidy frontend, add file (in my case a stream) to queue
4. Try to save queue as playlist

#### Expected behaviour:
- new playlist with file should be present in mopidy frontend section "Playlists"

#### Actual behaviour:
- mopidy frontend section "Playlists" is empty

#### Extra details:
- /var/log/mopidy/mopidy.log says: 2019-01-01 18:55:08,061 ERROR [2333:M3UBackend-4] mopidy.m3u.playlists: Error creating playlist [Radio Streams]: Permission denied
- /etc/mopidy/mopidy.conf says in section [m3u] playlists_dir = /mnt, which is owned by root and is not writable by user mopidy

#### My workaround (possible solution for dietpi setup scripts?!)
- mkdir /mnt/dietpi_userdata/mopidy/m3u, chmod/chown mopidy as owner (like the data and cache dir)
- nano /etc/mopidy/mopidy.conf, in section [m3u] playlists_dir = /mnt/dietpi_userdata/mopidy/m3u
- systemctl restart mopidy
- ... then I can add a playlist in mopidy frontend which is stored as .m3u8

**Comment by MichaIng** [comment]
@cyberlussi 
Thanks for your report.

**EDIT: In short: Your solution is just fine.**
You can also use `/mnt/dietpi_userdata/Music` for playlist, which is `dietpi` owned but writeable for `mopidy` user as well, which should be inside the `dietpi` group. Although have to check. We run: `useradd -rM mopidy -G dietpi,audio -s /usr/sbin/nologin`, but it looks like this user already exist after APT install, so this command should fail. `usermod mopidy -G dietpi,audio -s /usr/sbin/nologin` should be used then instead, or simply:` usermod -a -G dietpi mopidy`

Hmm jep we set media and playlist dirs to `/mnt` via config currently. However of course this is no valid media dir, I guess it is meant to be changed to wherever you like to place your media.

But a better start which is also writeable by the `mopidy` user is: `/mnt/dietpi_userdata/Music`
It is owned by user `dietpi`, but `mopidy` user is in dietpi group and the dir has group R/W permissions. This also matches the default setup of other media players we offer.
Also it does not match our online doc: https://dietpi.com/phpbb/viewtopic.php?f=8&t=5&start=80#p3611

I guess the idea behind setting those paths to `/mnt` is that `mopidyctl local scan` (respectively it's automatic background media scanner) justs scans locations under those configured dirs? So if you set your media+playlist dirs to `/mnt/dietpi_userdata/Music` it will not search/find media within e.g. your samba mount at `/mnt/samba`.

Just there are three dirs to configure:
```
[local]
media_dir = /mnt
[file]
media_dirs = /mnt
[m3u]
playlists_dir = /mnt
```
- local/media_dir: https://docs.mopidy.com/en/latest/ext/local/
  - Where it scans for local media files. There is a point in having this set to `/mnt`, so by default all mounts are scanned.
- file/media_dirs: https://docs.mopidy.com/en/latest/ext/file/
  - Where you can actually browse through via internal file browser. Not sure if/how it makes sense to have different dirs here and above.
- m3u/playlists_dir: https://docs.mopidy.com/en/latest/ext/m3u/
  - Where to read/write playlist files. Although one should be able to open those within local/media_dir and especially file/media_dirs as well.

All a bid confusing to me. Doesn't make much sense from my point of view to have mopidy scanning different locations for media files than having available for browsing. Having playlist files at a sub dir somehow has a point, although most users have playlist and music files within the same dir.
Further reading: https://discourse.mopidy.com/t/what-exactly-is-the-difference-between-extensions-local-and-file/1073
- local media dir needs to be scanned, but can be searched then.
- file media dirs don't need to be scanned (meta data will be imported on access), can be browsed, but cannot be searched.

Since playlists is the only thing that Mopidy need to write, I think having this set to `/mnt/dietpi_userdata/Music` makes most sense, while leaving the other two at `/mnt` is okay to have network and external drives included to scan/browse by default. This is usually world readable at least, so should work just fine to play Music from.
__________
PR opened to address the above: https://github.com/Fourdee/DietPi/pull/2385
- Playlists dir to `/mnt/dietpi_userdata/Music`
- Correctly add `mopidy` user (and systemd unit) to `dietpi` group, to assure R/W access to Music dir
- Preserve custom config by never overwriting any existing config file.

**Comment by cyberlussi** [comment]
Thank you for the extensive answer. So effectively it should work with my proposed workaround and also with yours. Using the already existent Music dir makes sense.
So I changed mopidy's m3u/playlists_dir to `/mnt/dietpi_userdata/Music`. But at least in my installation [Edit: 6.19.7] the group `dietpi` does **not** have `mopidy` as member: `getent group dietpi` gives `dietpi:x:1000:` and mopidy complains as before to have no permissions to write in `/mnt/dietpi_userdata/Music`. So I did ` usermod -a -G dietpi mopidy` as you suggested, then mopidy is able to write the m3u8 file after a service restart: `ls -la /mnt/dietpi_userdata/Music/` gives `-rw------- 1 mopidy audio       56 Jan  2 14:47 [Radio Streams].m3u8`

**Comment by MichaIng** [comment]
@cyberlussi 
Jep, we used `useradd` with option to define user groups and such, but the (new) APT install already creates the user, which is why we need to use `usermod`. I resolved that with the PR. I guess either this changed with moreless recent Mopidy versions or we used a different install method in the past (e.g. build from source), which did not include the user creation.

**Comment by Fourdee** [comment]
@MichaIng 

Thanks, great work, appears to be resolved in v6.20:
![image](https://user-images.githubusercontent.com/7077026/50697918-8743e780-103b-11e9-9ca4-47b345981c47.png)

Patch reinstall for existing installations?
